core/src/main/java/org/apache/iceberg/avro/AvroWriterBuilderFieldIdUtil.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.avro;
+
+import java.util.Deque;
+import org.apache.avro.Schema;
+
+/**
+ * This util class helps Avro DatumWriter builders to retrieve the correct field Id when building Avro DatumWriters
+ * with visitor pattern.
+ */
+public class AvroWriterBuilderFieldIdUtil {
```

Comment:
Minor: I think it would be a bit shorter to make this an abstract class that can be extended to get the current field ID via a protected method, but this is a good solution already.